### PR TITLE
Racheli

### PR DESCRIPTION
**Pull request by dkracheli**

### CODE REVIEW

File: middleware/authenticate.js [hunk 1]
```diff
@@ -1,32 +1,71 @@
...
-//     res.sendStatus(204)
-// })
-//
-// function generateAccessToken(user) {
-//     return jwt.sign(user, process.env.ACCESS_TOKEN_SECRET, {expiresIn: '15m'})
-// }
-//
-//
-// module.exports = {authJwt};
+const jwt = require('jsonwebtoken');
```
Comment (dkracheli) [inline middleware/authenticate.js:1]: @DaveAvigad

File: middleware/authenticate.js [hunk 2]
```diff
@@ -1,32 +1,71 @@
...
+    } else {
+      const user = await User.retrieve({ refreshToken: req.cookies.jwt });
+      req.user = user;
+      req.token = { refreshToken: req.body.refreshToken, accessToken: authHeader };
+      next();
+    }
+  });
+};
+
+module.exports = { generateToken, authenticateToken };
```
Comment (dkracheli) [inline middleware/authenticate.js:71]: @DaveAvigad

Comment (DaveAvigad) [reply middleware/authenticate.js:71]: @dkracheli row 52 is problematic? 
`console.log(req.cookies.jwt);`

File: controllers/loginController.js [hunk 3]
```diff
@@ -8,13 +8,18 @@ const loginControl = async (req, res, next) => {
   try {
     const user = await userExist(req.body.email);
     if (!user) throw new Error('user not exist');
+
     await validPassword(req.body.password, user.password);
```
Comment (DaveAvigad) [inline controllers/loginController.js:12]: validatePassword sound better no?

File: controllers/loginController.js [hunk 4]
```diff
@@ -30,7 +35,10 @@ const forgotPassControl = async (req, res) => {
...
     const hashedPassword = await bcrypt.hash(newPass, 12);
     await sendEmailPassword(newPass, user);
-    await User.update({ email: user.email }, { password: hashedPassword });
+    await User.update(
+      { email: user.email },
+      { password: hashedPassword },
+    );
 
     return res.status(200)
       .json({ message: 'A new password has been sent to the email' });
```
Comment (DaveAvigad) [inline controllers/loginController.js:44]: the controllers are too long. try to keep it no more than 3 lines. 💪

File: controllers/registerController.js [hunk 5]
```diff
@@ -33,6 +33,8 @@ const handleRegister = async (req, res) => {
 
 const handleConfirmCode = async (req, res) => {
   try {
+    const user = await userExist(req.body.email);
+    if (user) throw new Error('user already exist');
     const oneTimePassRecord = await existCode(req.body.email);
     await register.otpCompare(req.body.code, oneTimePassRecord.code);
     const accountID = await createAccount(req.body.email);
```
Comment (DaveAvigad) [inline controllers/registerController.js:40]: `accountId` keep the camleCase convention!

File: middleware/authenticate.js [hunk 6]
```diff
@@ -1,32 +1,71 @@
...
+
+function generateAccessToken(user) {
+  return jwt.sign(user, process.env.ACCESS_TOKEN_SECRET, { expiresIn: '5s' });
+}
+
+const refreshTokenVerify = async (req, res) => {
+  const refreshToken = req.cookies.jwt;
+  if (refreshToken == null) return res.status(403).json({ message: 'Unauthorized' });
+
+  const user = await User.retrieve({ refreshToken });
```
Comment (DaveAvigad) [inline middleware/authenticate.js:38]: `register.otpCompare(req.body.code, oneTimePassRecord.code);`
what if this throws?

File: middleware/authenticate.js [hunk 7]
```diff
@@ -1,32 +1,71 @@
...
+      if (res.statusCode !== 200) return res.send();
+      next();
+    } else {
+      const user = await User.retrieve({ refreshToken: req.cookies.jwt });
+      req.user = user;
+      req.token = { refreshToken: req.body.refreshToken, accessToken: authHeader };
+      next();
+    }
+  });
+};
```
Comment (DaveAvigad) [inline middleware/authenticate.js:69]: remember decoupling your code improves readability and modularity

File: middleware/authenticate.js (hunk 2 above)
Comment (DaveAvigad) [reply middleware/authenticate.js:71]: so you see the cookies on row 52 but do not get the header on row 54. I sent you a link on how to handle this flow appropriately. apparently, you need to send the token in the request body. 
[https://codeforgeek.com/refresh-token-jwt-nodejs-authentication/](url)

File: middleware/validator.js [hunk 8]
```diff
@@ -1,72 +1,4 @@
...
-      const user = await User.retrieve({ refreshToken: req.cookies.jwt });
-      req.user = user;
-      req.token = { refreshToken: req.body.refreshToken, accessToken: authHeader };
-      next();
-    }
-  });
-};
 
 const validation = (req, res, next) => {
   try {
```
Comment (DaveAvigad) [inline middleware/validator.js:4]: why empty?
if its still needs to be done, add `// note: todo`

**Review by DaveAvigad — COMMENTED** [review]
nice! I left some comments.
please try to shrink the handler's size and modular your code.

**Pull request merged**